### PR TITLE
Add SoundJson

### PR DESCRIPTION
**Pull request by huynhngocminh2511**

### CODE REVIEW

File: Zen/Configuration/UserConfiguration.cs [hunk 1]
```diff
@@ -1,32 +1,34 @@
...
 
         private UserConfiguration()
-        {
-            var sharedPreferences = PreferenceManager.GetDefaultSharedPreferences(Application.Context);
-            var configurationText = sharedPreferences.GetString(PreferenceKey, "");
-
-            Configuration = JsonConvert.DeserializeObject<Configuration>(configurationText) 
-                                       ?? new Configuration();
+        {
+            Save();
```
Comment (voquanghoa) [inline Zen/Configuration/UserConfiguration.cs]: I think it need to load at first, not save.

File: Zen/Configuration/UserConfiguration.cs [hunk 2]
```diff
@@ -1,32 +1,34 @@
...
             var sharedPreferencesEditor = PreferenceManager.GetDefaultSharedPreferences(Application.Context).Edit();
             var configJson = JsonConvert.SerializeObject(Configuration);
-            sharedPreferencesEditor.PutString(PreferenceKey, configJson);
+            sharedPreferencesEditor.PutString("UserConfiguration", configJson);
             sharedPreferencesEditor.Apply();
         }
+
+        public void Load()
+        {
+            var sharedPreferences = PreferenceManager.GetDefaultSharedPreferences(Application.Context);
```
Comment (voquanghoa) [inline Zen/Configuration/UserConfiguration.cs]: Why did you remove the constant?

File: Zen/Configuration/UserConfiguration.cs [hunk 3]
```diff
@@ -1,32 +1,34 @@
...
+            Configuration = JsonConvert.DeserializeObject<Configuration>(sharedPreferences.GetString("UserConfiguration", ""));
```
Comment (voquanghoa) [inline Zen/Configuration/UserConfiguration.cs]: Handle when it is `null`.

File: Zen/Controllers/SoundsController.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        private const string ambientSound = "ambient";
+
+        private List<Sound> sounds;
+
+        private Configuration.Configuration configuration = UserConfiguration.Instance.Configuration;
+
+        public void LoadConfiguration()
+        {
+            configuration = UserConfiguration.Instance.Configuration;
```
Comment (voquanghoa) [inline Zen/Controllers/SoundsController.cs]: Because you have already have this
```C#
private Configuration.Configuration configuration = UserConfiguration.Instance.Configuration;
```
So this line 
```C#
configuration = UserConfiguration.Instance.Configuration;
```
Is redundant.

File: Zen/Controllers/SoundsController.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        private const string backgroundSound = "background";
+
+        private const string ambientSound = "ambient";
+
+        private List<Sound> sounds;
+
+        private Configuration.Configuration configuration = UserConfiguration.Instance.Configuration;
+
+        public void LoadConfiguration()
```
Comment (voquanghoa) [inline Zen/Controllers/SoundsController.cs]: For Singleton class, you should make the `contructor` is `private` and remove this method

File: Zen/Controllers/SoundsController.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                LoadSounds(context);
+            }
+
+            return sounds;
+        }
+
+        public Sound GetSoundByName(Context context, string fileName)
+        {
+            if (sounds == null)
```
Comment (voquanghoa) [inline Zen/Controllers/SoundsController.cs]: For singleton class, you should make sure the data is ready every time we need it.

File: Zen/Controllers/SoundsController.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                LoadSounds(context);
+            }
+
+            return sounds.Where(x => x.Type.Equals(type)).ToList();
+        }
+
+        public Sound BellSelected { private set { } get => configuration.Bell; }
+
+        public Sound BackgroundSoundSelected { private set { } get => configuration.BackgroundSound; }
```
Comment (voquanghoa) [inline Zen/Controllers/SoundsController.cs]: Remove `private set{}`. It no useful

File: Zen/Fragments/GuideFragment.cs [hunk 8]
```diff
@@ -1,45 +1,24 @@
...
     public class GuideFragment : Fragment
     {
-        [InjectView(Resource.Id.imgGuide)] private ImageView imgGuide;
-
-        [InjectView(Resource.Id.rvGuideLevels)] private RecyclerView rvGuideLevels;
-
-        private GuideLevelsAdapter adapter;
-
-        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
+        public override void OnCreate(Bundle savedInstanceState)
```
Comment (voquanghoa) [inline Zen/Fragments/GuideFragment.cs]: Remove empty overwrite

File: Zen/Fragments/HomeFragment.cs [hunk 9]
```diff
@@ -24,7 +24,7 @@ public class HomeFragment : Fragment
 
         public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
         {
-            view = isHomeStart ? inflater.Inflate(Resource.Layout.HomeLayout, container, false) : view = inflater.Inflate(Resource.Layout.HomeResetLayout, container, false);
+            view = IsHomeStart ? inflater.Inflate(Resource.Layout.HomeLayout, container, false) : view = inflater.Inflate(Resource.Layout.HomeResetLayout, container, false);
```
Comment (voquanghoa) [inline Zen/Fragments/HomeFragment.cs]: Seems it never false

File: Zen/Fragments/SettingFragment.cs [hunk 10]
```diff
@@ -67,5 +71,14 @@ private static void SetWidgets(TextView textView, bool status, string value)
         {
             textView.Text = !status ? "Off" : $"On / {value}";
         }
+
+        private static void SetWidgets(TextView textView, bool status, List<Sound> value)
+        {
+            var values = "";
```
Comment (voquanghoa) [inline Zen/Fragments/SettingFragment.cs]: You can do like this
```C#
var values = value.Aggregate((x,y) => $"{a} {b}");
```

File: Zen/Models/AmbientSoundTypes.cs [hunk 11]
```diff
@@ -10,56 +12,57 @@ public static class AmbientSoundTypes
...
         {
-            AmbientSoundTypesList.ForEach(x => x.Selected = x.Name == typeName);
+            AmbientSoundTypesList.ForEach(x => x.Selected = AmbientSounds.Any(y => y.Id == x.Id));
         }
 
-        public static string GetChoosingType()
+        public static List<Sound> GetChoosingType(Context context)
         {
-            return AmbientSoundTypesList.First(x => x.Selected).Name;
+            return AmbientSoundTypesList.Where(x => x.Selected == true)
```
Comment (voquanghoa) [inline Zen/Models/AmbientSoundTypes.cs]:
```C#
.Where(x => x.Selected)
```

File: Zen/Models/Sound.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        [JsonProperty("name")]
+        public string Name { get; set; }
+
+        [JsonProperty("url")]
+        public string Url { get; set; }
+
+        [JsonProperty("type")]
+        public string Type { get; set; }
+
+        public MediaPlayer SoundMediaPlayer { set; get; } = new MediaPlayer();
```
Comment (voquanghoa) [inline Zen/Models/Sound.cs]: Don't put `MediaPlayer` here

File: Zen/Ultilities/AlertDialogUltility.cs [hunk 13]
```diff
@@ -24,5 +27,32 @@ public static void ShowDialog(List<string> data, TextView textView, string title
             alertDialog.SetView(listView);
             alertDialog.Show();
         }
+
+        public static void ShowDialog(List<Sound> data, TextView textView, string title)
```
Comment (voquanghoa) [inline Zen/Ultilities/AlertDialogUltility.cs]: Rename `data` --> `sounds`

File: Zen/Ultilities/AlertDialogUltility.cs [hunk 14]
```diff
@@ -24,5 +27,32 @@ public static void ShowDialog(List<string> data, TextView textView, string title
...
+        {
+            var builder = new AlertDialog.Builder(textView.Context);
```
Comment (voquanghoa) [inline Zen/Ultilities/AlertDialogUltility.cs]:
```C#
var alertDialog = new AlertDialog.Builder(textView.Context).SetTitle(title).Create();
```

File: Zen/Ultilities/AlertDialogUltility.cs (hunk 13 above)
Comment (voquanghoa) [inline Zen/Ultilities/AlertDialogUltility.cs]: `ShowDialog` is not clear --> `ShowSoundSelectionDialog`
`textView` is not clear --> `outputTextView`

File: Zen/Fragments/SettingFragment.cs [hunk 15]
```diff
@@ -67,5 +71,14 @@ private static void SetWidgets(TextView textView, bool status, string value)
...
+
+            value.ForEach(x => values = values + " " + x.Name);
+
+            textView.Text = !status ? "Off" : $"On / {values}";
```
Comment (voquanghoa) [inline Zen/Fragments/SettingFragment.cs]: Invert status. And extract this to a function.

File: Zen/Fragments/SettingFragment.cs [hunk 16]
```diff
@@ -67,5 +71,14 @@ private static void SetWidgets(TextView textView, bool status, string value)
         {
             textView.Text = !status ? "Off" : $"On / {value}";
```
Comment (voquanghoa) [inline Zen/Fragments/SettingFragment.cs]: Duplicated code

File: Zen/Controllers/SoundsController.cs [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        public void LoadConfiguration()
+        {
+            configuration = UserConfiguration.Instance.Configuration;
+        }
+
+        private void LoadSounds(Context context)
+        {
+            var content = "";
+
+            using (var sr = new StreamReader(context.Assets.Open("Sounds.json")))
```
Comment (voquanghoa) [inline Zen/Controllers/SoundsController.cs]: Make it shorter
```C#
using (var sr = new StreamReader(context.Assets.Open("Sounds.json")))
            {
               var content = sr.ReadToEnd();
               sounds = JsonConvert.DeserializeObject<List<Sound>>(content);
           }
```

File: Zen/Models/AmbientSoundTypes.cs [hunk 18]
```diff
@@ -1,7 +1,9 @@
...
+﻿using Android.Content;
+using System.Collections.Generic;
 using System.Linq;
 using Xamarin.Forms.Internals;
-
+using Zen.Controllers;
+
 namespace Zen.Models
 {
     public static class AmbientSoundTypes
```
Comment (voquanghoa) [inline Zen/Models/AmbientSoundTypes.cs]: Can we remove this file because we have the file `Sounds.json`?

File: Zen/Activities/SettingMusicActivity.cs [hunk 19]
```diff
@@ -34,7 +35,7 @@ private void Back(object sender, EventArgs e)
         [InjectOnClick(Resource.Id.cvSong)]
         private void ChooseSong(object sender, EventArgs e)
         {
-            AlertDialogUltility.ShowDialog(Resources.GetStringArray(Resource.Array.songs).ToList(), tvSong, "Choose song");
+            AlertDialogUltility.ShowDialog(SoundsController.Instance.GetBackgroundSounds(this), tvSong, "Choose song");
```
Comment (voquanghoa) [inline Zen/Activities/SettingMusicActivity.cs]: Lines is too long, please split it

File: Zen/Adapters/TypesAdapter.cs [hunk 20]
```diff
@@ -34,15 +34,10 @@ public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int
...
-            if (types[position].Selected) return;
-            foreach (var t in types)
-            {
-                if (t != types[position])
-                {
-                    t.Selected = false;
-                }
-            }
-            types[position].Selected = true;
+            types[position].Selected = types[position].Selected == true ? false : true;
```
Comment (voquanghoa) [inline Zen/Adapters/TypesAdapter.cs]: How about
```C#
types[position].Selected = !types[position].Selected;

if(!types.Any(x => x.Selected))
{
   types[position].Selected = true;
}
```

File: Zen/Activities/HomeListenActivity.cs [hunk 21]
```diff
@@ -25,12 +31,62 @@ public class HomeListenActivity : Activity
...
+        }
+
+        private void InitSounds()
+        {
+            SoundsController.Instance.LoadConfiguration();
+            soundsController = SoundsController.Instance;
+
+            configuration = UserConfiguration.Instance.Configuration;
+
+            if (configuration.BellEnable == true) PlaySoundIntervalTime(soundsController.BellSelected);
```
Comment (voquanghoa) [inline Zen/Activities/HomeListenActivity.cs]: Remove `== true`

File: Zen/Activities/HomeListenActivity.cs [hunk 22]
```diff
@@ -25,12 +31,62 @@ public class HomeListenActivity : Activity
...
+
+        private void PlaySoundIntervalTime(Sound sound)
+        {
+            var task = Task.Run(() =>
+            {
+                sound.SoundMediaPlayer.SetDataSource(Assets.OpenFd(sound.Url));
+                sound.SoundMediaPlayer.Prepare();
+                sound.SoundMediaPlayer.Start();
+
+                while (true)
```
Comment (voquanghoa) [inline Zen/Activities/HomeListenActivity.cs]: Need to stop this

File: .gitignore [hunk 23]
```diff
@@ -23,7 +23,7 @@ bld/
 [Bb]in/
 [Oo]bj/
 [Ll]og/
-Zen/Resources/Resource.Designer.cs
```
Comment (voquanghoa) [inline .gitignore]: Why did you remove this line.

File: Zen/Adapters/TypesAdapter.cs (hunk 20 above)
Comment (huynhngocminh2511) [reply Zen/Adapters/TypesAdapter.cs]: Done

File: Zen/Activities/HomeListenActivity.cs (hunk 21 above)
Comment (huynhngocminh2511) [reply Zen/Activities/HomeListenActivity.cs]: Done

File: Zen/Configuration/UserConfiguration.cs (hunk 1 above)
Comment (huynhngocminh2511) [reply Zen/Configuration/UserConfiguration.cs]: Done

File: Zen/Configuration/UserConfiguration.cs (hunk 3 above)
Comment (huynhngocminh2511) [reply Zen/Configuration/UserConfiguration.cs]: Done

File: Zen/Controllers/SoundsController.cs (hunk 5 above)
Comment (huynhngocminh2511) [reply Zen/Controllers/SoundsController.cs]: Done

Comment (huynhngocminh2511) [reply Zen/Controllers/SoundsController.cs]: Done

File: Zen/Activities/SettingMusicActivity.cs (hunk 19 above)
Comment (huynhngocminh2511) [reply Zen/Activities/SettingMusicActivity.cs]: Done

File: Zen/Controllers/SoundsController.cs (hunk 6 above)
Comment (huynhngocminh2511) [reply Zen/Controllers/SoundsController.cs]: Done

File: Zen/Controllers/SoundsController.cs (hunk 7 above)
Comment (huynhngocminh2511) [reply Zen/Controllers/SoundsController.cs]: Done

File: Zen/Models/Sound.cs (hunk 12 above)
Comment (huynhngocminh2511) [reply Zen/Models/Sound.cs]: Done

File: Zen/Ultilities/AlertDialogUltility.cs (hunk 13 above)
Comment (huynhngocminh2511) [reply Zen/Ultilities/AlertDialogUltility.cs]: Done

File: Zen/Ultilities/AlertDialogUltility.cs (hunk 14 above)
Comment (huynhngocminh2511) [reply Zen/Ultilities/AlertDialogUltility.cs]: Done

File: Zen/Ultilities/AlertDialogUltility.cs (hunk 13 above)
Comment (huynhngocminh2511) [reply Zen/Ultilities/AlertDialogUltility.cs]: Done

File: Zen/Fragments/SettingFragment.cs (hunk 15 above)
Comment (huynhngocminh2511) [reply Zen/Fragments/SettingFragment.cs]: Done

File: Zen/Fragments/SettingFragment.cs (hunk 16 above)
Comment (huynhngocminh2511) [reply Zen/Fragments/SettingFragment.cs]: Done

File: Zen/Fragments/SettingFragment.cs (hunk 10 above)
Comment (huynhngocminh2511) [reply Zen/Fragments/SettingFragment.cs]: Done

File: Zen/Activities/HomeListenActivity.cs (hunk 22 above)
Comment (huynhngocminh2511) [reply Zen/Activities/HomeListenActivity.cs]: Done

File: Zen/Controllers/SoundsController.cs (hunk 4 above)
Comment (voquanghoa) [reply Zen/Controllers/SoundsController.cs]: Still redundant

File: Zen/Controllers/SoundsController.cs (hunk 17 above)
Comment (voquanghoa) [reply Zen/Controllers/SoundsController.cs]: ???
